src/response.rs

```diff
@@ -317,28 +325,32 @@ impl Response {
             (true, _) => {
                 debug!("Chunked body in response");
                 Box::new(PoolReturnRead::new(
-                    &unit.agent,
-                    &unit.url,
                     ChunkDecoder::new(stream),
                 ))
             }
             // Responses with a content-length header means we should limit the reading
             // of the body to the number of bytes in the header. Once done, we can
             // return the underlying stream to the connection pool.
+            (false, Some(0)) => {
+                error!("received a zero-length body, when this should have been handled earlier. connection will drop.");
+                Box::new(std::io::empty())
+            }
```

Comment:
I agree in principle that we should panic when (a) it's a violation of internal assumptions in the library, and (b) we can't recover from it. However, in this case it's quite easy to recover - we just return an empty body, and that's correct. And since we'd really like to never have a user-facing panic, this buys us a little safety in case we are wrong about the internal assumptions.

I feel a little bad about that `Result`, but `stream_to_reader`'s only caller is `do_from_stream`, which already returns a result. Really, `stream_to_reader` could be considered as part of `do_from_stream`. In fact, I kind of have in mind a refactoring that would move the analysis of Content-Length and Transfer-Encoding header fields up a level, into do_from_stream. I think this would allow us to get rid of the awkward thing in do_from_stream where we construct a Response with an empty reader, and then immediately replace that reader with `response.stream_to_reader(...)`.

Reply:
> I agree in principle that we should panic when (a) it's a violation of internal assumptions in the library, and (b) we can't recover from it. However, in this case it's quite easy to recover

Not sure that the ease of which we can recover from a situation has any bearing – either the lib does what it should, or it doesn't. If it doesn't, I can see bad consequences from an unexpected internal state, like request smuggling or just plain garbage in a reused stream. Not saying that either of those would happen here, but it's the principle of not "plastering over" badness simply to avoid a panic.